# CYCLOPS 2.0
CYCLOPS (Cyclic Ordering by Periodic Structure) is a tool designed to reconstruct the temporal ordering of high dimensional data that is generated by a common periodic process. This module is an improvement to the original CYCLOPS. The improved CYCLOPS 2.0 is written for Julia 1.6.  
  
## CYCLOPS.jl
This module contains all the code necessary to pre-process data, train CYCLOPS 2.0, and calculate cosinors of best fit for all transcripts using CYCLOPS sample phase predictions.
  
# How to Run CYCLOPS
A description of how to use CYCLOPS 2.0 to order your own dataset. Outlined are the required packages, expression file format, input arguments, and hyper parameters. The CYCLOPS.jl script is copy and pasted into a terminal running Julia 1.6.

## Contents
1. Packages
2. Expression Data File
3. Seed Genes
4. Covariates
5. Hyper Parameters*
6. Sample Collection Times
7. CYCLOPS.Fit
8. CYCLOPS.Align
9. Contact
  
## 1. Packages
This module requires the following pacakges (- version):  
```
CSV --------------- v0.10.4  
DataFrames -------- v1.3.4  
Distributions ----- v0.25.68  
Flux -------------- v0.13.5  
MultipleTesting --- v0.5.1  
MultivariateStats - v0.10.0  
Plots ------------- v1.38.11  
PyPlot ------------ v2.11.0  
Revise ------------ v3.4.0  
StatsBase --------- v0.33.21  
XLSX -------------- v0.8.4  
Dates  
Distributed  
LinearAlgebra  
Random  
Statistics  
```
  
## 2. Expression Data File
The expression data are a required input to the *CYCLOPS.Fit* and *CYCLOPS.Align* functions. The format of the expression data file is as follows:  
  
1. Each column is a sample.
2. Each row is a transcript.
3. The first column of the dataset contains gene symbols and covariate labels (see 'Covariates' and 'Hyper Parameters').
4. The first 'k' rows contain covariates and all following rows contain numeric expression data (see 'Covariates').
5. Samples and columns with 'missing' or 'NaN' values should be removed from the dataset.
6. Column names should start with letters and should only contain numbers, letters, and underscores ('_').
7. Gene symbols should start with letters.
8. Duplicate gene symbols are allowed.
9. Duplicate column names are **NOT** allowed.
  
### Example Expression Data File
```
12869×653 DataFrame
   Row │ Gene_Symbol   GTEX_1117F_2826_SM_5GZXL  GTEX_1122O_1226_SM_5H113  GTEX ⋯
───────┼─────────────────────────────────────────────────────────────────────────
     1 │ tissueType_D  NonTumor                  Tumor                     NonT ⋯
     2 │ site_D        B1                        B1                        B1
     3 │ WASH7P        10.04                     3.248                     4.82
     4 │ LINC01128     5.357                     7.199                     4.57
     5 │ SAMD11        0.6739                    1.213                     0.46 ⋯
     6 │ NOC2L         64.54                     62.24                     73.7
     7 | NOC2L         65.52                     61.18                     74.8
   ⋮   |      ⋮                   ⋮                         ⋮                   ⋱
 12865 │ MTCP1         6.513                     6.308                     7.13
 12866 │ BRCC3         9.685                     10.12                     16.1
 12867 │ VBP1          34.65                     30.77                     25.4 ⋯
 12868 │ CLIC2         21.85                     30.05                     16.8
 12869 │ TMLHE         5.04                      4.06                      5.24
```
  
## 3. Seed Genes  
The seed genes are a required input to the *CYCLOPS.Fit* function. Seed genes must be provided as a vector of strings (not symbols). Also consider:  
  
1. Case matters!  
   *"Acot4" in the seed gene list will not match "ACOT4" in the gene symbols of the expression data.*  
2. Provide enough seed genes.  
   *The number of seed genes should be greater than ':eigen_max' (see 'Hyper Parameters').*  
3. Seed genes start with letters.  
   *Gene symbols in the expression data should start with letters, therefore seed genes should also start with letters.*  
  
### Example Seed Genes
```
71-element Vector{String}:
 "ACOT4"
 "ACSM5"
 "ADORA1"
 "ADRB3"
 "ALAS1"
 "ANGPTL2"
 "ARHGAP20"
 "ARNTL"
 ⋮
 "TP53INP2"
 "TSC22D3"
 "TSPAN4"
 "TUSC5"
 "USP2"
 "WEE1"
 "ZEB2"
```
  
## 4. Covariates  
The expression data may contain rows of grouping variables (discontinuous covariates) or continuous variables (continuous covariates). The following constraints apply:  
  
1. All covariate rows must be above expression data.
2. Grouping variables (discontinous covariates) must start with letters.  
3. Continuous variables (continuous covariates) must **only** contain numbers.  
4. Within a row of grouping variables (discontinuous covariates), each group must contain at least two (2) or more samples.  
   *Consider 'tissue type' as a covariate. If the data has 'NonTumor' and 'Tumor' samples, there should be at least two (2) 'NonTumor' and two (2) 'Tumor' samples in the data. Ideally, the number of samples in each group is greater than ':eigen_max' (see 'Hyper Parameters').*  
5. All samples must have have values for **all** covariates (rows).  
6. Covariate rows must have unique names.  
7. Covariate rows have regex identifiers in the gene symbol column.  
   *The example below, discontinuous covariates end in '_D' and continuous covariates end in '_C'. See 'Hyper Parameters.'*  
  
### Example Covariates
```
   Row │ Gene_Symbol   GTEX_1117F_2826_SM_5GZXL  GTEX_1122O_1226_SM_5H113  GTEX ⋯
───────┼─────────────────────────────────────────────────────────────────────────
     1 │ tissueType_D  NonTumor                  Tumor                     NonT ⋯
     2 │ site_D        B1                        B1                        B1
     3 │ age_C         25                        56                        62
     4 │ GENE1         4.32                      27.63                     18.43
```

## 5. Hyper Parameters  
Hyper parameters are stored in a single dictionary in order to reduce the number of individual input arguments to the *CYCLOPS.Fit* and *CYCLOPS.Align* functions. Below is the default hyper parameter dictionary, including default values. It is not recommended to alter default values unless absolutely necessary.
  
```julia
Dict(  
  :regex_cont => r".*_C",                # What is the regex match for continuous covariates in the data file?
  :regex_disc => r".*_D",                # What is the regex match for discontinuous covariates in the data file?

  :blunt_percent => 0.975,               # What is the percentile cutoff below (lower) and above (upper) which values are capped?

  :seed_min_CV => 0.14,                  # The minimum coefficient of variation a gene of interest may have to be included in eigen gene transformation
  :seed_max_CV => 0.7,                   # The maximum coefficient of a variation a gene of interest may have to be included in eigen gene transformation
  :seed_mth_Gene => 10000,               # The minimum mean a gene of interest may have to be included in eigen gene transformation

  :norm_gene_level => true,              # Does mean normalization occur at the seed gene level
  :norm_disc => false,                   # Does batch mean normalization occur at the seed gene level
  :norm_disc_cov => 1,                   # Which discontinuous covariate is used to mean normalize seed level data

  :eigen_reg => true,                    # Does regression against a covariate occur at the eigen gene level
  :eigen_reg_disc_cov => 1,              # Which discontinous covariate is used for regression
  :eigen_reg_exclude => false,           # Are eigen genes with r squared greater than cutoff removed from final eigen data output
  :eigen_reg_r_squared_cutoff => 0.6,    # This cutoff is used to determine whether an eigen gene is excluded from final eigen data used for training
  :eigen_reg_remove_correct => false,    # Is the first eigen gene removed (true --> default) or it's contributed variance corrected by batch regression (false)

  :eigen_first_var => false,             # Is a captured variance cutoff on the first eigen gene used
  :eigen_first_var_cutoff => 0.85,       # Cutoff used on captured variance of first eigen gene

  :eigen_total_var => 0.85,              # Minimum amount of variance required to be captured by included dimensions of eigen gene data
  :eigen_contr_var => 0.06,              # Minimum amount of variance required to be captured by a single dimension of eigen gene data
  :eigen_var_override => false,          # Is the minimum amount of contributed variance ignored
  :eigen_max => 30,                      # Maximum number of dimensions allowed to be kept in eigen gene data

  :out_covariates => true,               # Are covariates included in eigen gene data
  :out_use_disc_cov => true,             # Are discontinuous covariates included in eigen gene data
  :out_all_disc_cov => true,             # Are all discontinuous covariates included if included in eigen gene data
  :out_disc_cov => 1,                    # Which discontinuous covariates are included at the bottom of the eigen gene data, if not all discontinuous covariates
  :out_use_cont_cov => false,            # Are continuous covariates included in eigen data
  :out_all_cont_cov => true,             # Are all continuous covariates included in eigen gene data
  :out_use_norm_cont_cov => false,       # Are continuous covariates Normalized
  :out_all_norm_cont_cov => true,        # Are all continuous covariates normalized
  :out_cont_cov => 1,                    # Which continuous covariates are included at the bottom of the eigen gene data, if not all continuous covariates, or which continuous covariates are normalized if not all
  :out_norm_cont_cov => 1,               # Which continuous covariates are normalized if not all continuous covariates are included, and only specific ones are included

  :init_scale_change => true,            # Are scales changed
  :init_scale_1 => false,                # Are all scales initialized such that the model sees them all as having scale 1
                                         # Or they'll be initilized halfway between 1 and their regression estimate.

  :train_n_models => 80,                 # How many models are being trained
  :train_μA => 0.001,                    # Learning rate of ADAM optimizer
  :train_β => (0.9, 0.999),              # β parameter for ADAM optimizer
  :train_min_steps => 1500,              # Minimum number of training steps per model
  :train_max_steps => 2050,              # Maximum number of training steps per model
  :train_μA_scale_lim => 1000,           # Factor used to divide learning rate to establish smallest the learning rate may shrink to
  :train_circular => false,              # Train symmetrically
  :train_collection_times => true,       # Train using known times
  :train_collection_time_balance => 0.1, # How is the true time loss rescaled

  :cosine_shift_iterations => 192,       # How many different shifts are tried to find the ideal shift
  :cosine_covariate_offset => true,      # Are offsets calculated by covariates

  :align_p_cutoff => 0.05,               # When aligning the acrophases, what genes are included according to the specified p-cutoff
  :align_base => "radians",              # What is the base of the list (:align_acrophases or :align_phases)? "radians" or "hours"
  :align_disc => false,                  # Is a discontinuous covariate used to align (true or false)
  :align_disc_cov => 1,                  # Which discontinuous covariate is used to choose samples to separately align (is an integer)
  :align_other_covariates => false,      # Are other covariates included
  :align_batch_only => false,

  :X_Val_k => 10,                        # How many folds used in cross validation.
  :X_Val_omit_size => 0.1)               # What is the fraction of samples left out per fold
```
  
There are also parameters that do not have default values. These parameters include:  
```julia
:train_sample_id    # Array{Symbol, 1}. Sample ids for which collection times exist.  
:train_sample_phase # Array{Number, 1}. Collection times for each sample id. 'train_sample_id' and 'train_sample_phase' must be the same length.
  
:align_genes        # Array{String, 1}. Genes with known acrophases.  
:align_acrophases   # Array{Number, 1}. Acrophases for each gene. 'align_genes' and 'align_acrophases' must be the same length.
  
:align_samples      # Array{String, 1}. Sample ids for which collection times exist.  
:align_phases       # Array{Number, 1}. Collection times for each sample id. 'align_samples' and 'align_phases' must be the same length.
```

## 6. Sample Collection Times  
Sample collection times may be provided and added to the hyper parameter dictionary. These may be used in two different ways:
  
1. Semi-supervised training.  
   *CYCLOPS optimizes the distance between predicted and provided sample collection times **while training** when sample ids and collection times are added via 'train_sample_id' and 'train_sample_phase'.*  
2. Post-training alignment.  
   *CYCLOPS predicted phases are aligned to sample collection times **after training** when added via 'align_samples' and 'algin_phases'.*  

## 7. CYCLOPS.Fit  
CYCLOPS.Fit has three (3) input argument, and five (5) outputs. Input arguments (in order) are:
  
1. the expression data (as described in section 2),  
2. the seed genes (as described in section 3),  
3. and the hyper parameter dictionary (as described in section 5).  
  
Outputs (in order) are:  
  
1. the eigen data,  
2. the sample fit,  
3. the eigen gene correlations,  
4. the trained model,  
5. and the updated hyper parameter dictionary.  
  
### Example Usage of CYCLOPS.Fit  
  
```julia
eigendata, samplefit, eigendatacorrelations, trainedmodel, updatedparameters = CYCLOPS.Fit(expressiondata, seedgenes, hyperparameters)
```

## 8. CYCLOPS.Align  
CYCLOPS.Align has six (6) input arguments, and saves files to a directory (no outputs). Input arguments (in order) are:  
  
1. the expression data (as described in section 2),  
2. the sample fit (second output of CYCLOPS.Fit),  
3. the eigen gene correlations (third output of CYCLOPS.Fit),  
4. the trained model (fourth output of CYCLOPS.Fit),  
5. the updated hyper parameter dictionary (fifth output of CYCLOPS.Fit),  
6. and the output path where results are saved.  
  
### Example Usage of CYCLOPS.Align  
  
```julia
CYCLOPS.Align(expressiondata, samplefit, eigendatacorrelations, trainedmodel, updatedparameters, outputpath)
```

## 9. Contact
Please contact janham@pennmedicine.upenn.edu with questions. Kindly make the subject of the email '***GitHub CYCLOPS 2.0***'. Please copy paste the full error in the body of the email, and provide the versions of all required packages. Further, provide the script and all necessary files to recreate the error you are encountering. Good luck and happy ordering.
